zeppelin_converter.py

```diff
@@ -0,0 +1,200 @@
+
+        def gen_table(m):
+            rows = m['data'].split('\n')
+            if rows != []:
+                header_row = rows[0]
+                body_rows = rows[1:]
+                create_md_row(header_row, True)
+                for row in body_rows:
+                    create_md_row(row)
+
+        """
+        Output options routes Zeppelin output types to corresponding
+        functions for it to be handled. To add support for other output
+        types, add the file type to the dictionary and create the necessary
+        function to handle it.
+
+
+        """
+        output_options = {
+            'HTML': gen_image,
+            'TEXT': gen_text,
+            'TABLE': gen_table
+        }
+
+        def process_output(p):
+            if 'editorMode' in p['config']:
+                mode = p['config']['editorMode'].split('/')[-1]
+                if p['results']['msg'] != []:
+                    msg = p['results']['msg'][0]
+                    if mode != 'text' and mode != 'markdown':
+                        output_options[msg['type']](msg)
+
+        def parse_date(d):
+            return datetime.strptime(d, '%b %d, %Y %I:%M:%S %p')
+
+        def process_date_created(t):
+            if self.date_created == 'N/A':
+                self.date_created = t
+            if parse_date(t) < parse_date(self.date_created):
+                self.date_created = t
+
+        def process_date_updated(t):
+            if self.date_updated == 'N/A':
+                self.date_updated = t
+            if parse_date(t) > parse_date(self.date_updated):
+                self.date_updated = t
+
+        for paragraph in t['paragraphs']:
+            if 'user' in paragraph.keys():
+                self.user = paragraph['user']
+            if 'dateCreated' in paragraph.keys():
+                process_date_created(paragraph['dateCreated'])
+            if 'dateUpdated' in paragraph.keys():
+                process_date_updated(paragraph['dateUpdated'])
+            if 'text' in paragraph.keys():
+                process_input(paragraph['text'])
+            if 'results' in paragraph.keys():
+                process_output(paragraph)
+
+    """
+    Takes in a .json file as input and convert it to Markdown format,
+    saving the generated .png images into ./images.
+
+    """
+    def convert(self):
+        if len(sys.argv) < 2:
+            raise ValueError('ERROR: No input file specified')
+        elif len(sys.argv) > 2:
+            raise ValueError('ERROR: Too many arguments')
+
+        f = sys.argv[1]
+
+        try:
+            with open(f, 'rb') as raw:
+                t = json.load(raw)
+                with open('knowledge.md', 'w') as fout:
+                    self.gen_body(t)        # Step 1: create the body
+                    self.gen_header(t)      # Step 2: create the md header
+                    fout.write(self.out)    # Step 3: write to file
+
+        except ValueError as err:
+            print err.args
```

Comment:
In case something goes wrong the program should exit with return code == 1. If you catch this exception the program will have an exit code == 0 . See https://docs.python.org/3/library/sys.html#sys.exit